File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidTree(root, null, null);        
    }

    public boolean isValidTree(TreeNode currNode, TreeNode minNode, TreeNode maxNode){

        if(currNode == null){
            return true;
        }

        if(minNode != null && minNode.val >= currNode.val){
            return false;
        }
        if(maxNode != null && maxNode.val <= currNode.val){
            return false;
        }

        return isValidTree(currNode.left, minNode, currNode) && isValidTree(currNode.right, currNode, maxNode);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def isBSTValid(root, minNode, maxNode):

            if root is None:
                return True
            
            if minNode and minNode.val >= root.val:
                return False
            elif maxNode and maxNode.val <= root.val:
                return False

            return isBSTValid(root.left, minNode, root) and isBSTValid(root.right, root, maxNode)

        return isBSTValid(root, None, None)